### PR TITLE
Programmatic Favicon notifications

### PR DESCRIPTION
**Pull request by mikedevelops**
A service that will update a users favicon.

I've added a "preset" method to the public API called `addCircleNotification(color, size)` which will cover most cases, as well as a `addCustomNotification((canvas, ctx) => {})` to allow users of the service to manipulate the favicon as they wish, however I suspect we'll end up expanding on the "preset" methods to cover more common cases.

The service "gracefully degrades", therefore environments that do not support any of it's features will silently fail with no adverse effect on the UX.

screen shot 2018-01-25 at 17 07 41

### CODE REVIEW

File: js/main.js [hunk 1]
```diff
@@ -75,6 +76,28 @@
             supported: !lt10,
             showInputNode: lt10
         });
+
+        // Notifications
+        pulsar.faviconEditor.init();
+
+        // NOTIFICATIONS DEBUG, PLS REMOVE, THX
```
Comment (Stanton) [inline js/main.js]: To keep in Lexicon or to remove as per comment?

File: .babelrc [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-  "presets": ["es2015"]
+  "presets": ["es2015", "es2017"]
```
Comment (asmblah) [inline .babelrc]: I know we discussed it before, but as you're changing this anyway, should we use the `env` preset now?

File: js/Notifications/FaviconEditor.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        let data;
+
+        this.favicons.forEach(({ node }) => {
+            if (filterFileExtension(node.href, 'ico png') || filterDataEncodedURI(node.href)) {
+                setupPromises.push(this.setup(node));
+            }
+        });
+
+        try {
+            favicons = await Promise.all(setupPromises);
```
Comment (asmblah) [inline js/Notifications/FaviconEditor.js]: Nice 👍

File: tests/js/web/Notifications/FaviconEditorTest.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
+const FaviconEditor = require('../../../../js/Notifications/FaviconEditor');
+const $ = require('jquery');
+const vanillaFavicon = require('./snapshots/favicon');
+const redCircleFavicon = require('./snapshots/faviconWithCricleNotification?radius=10&color=red');
```
Comment (asmblah) [inline tests/js/web/Notifications/FaviconEditorTest.js]: `faviconWithCricle` -> `faviconWithCircle` (and the file to match of course) 😄

File: tests/js/web/Notifications/FaviconEditorTest.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                    .catch(done);
+            });
+        });
+
+        describe('addCustomNotification', () => {
+            it('should draw a custom graphic onto an image', (done) => {
+                faviconEditor = new FaviconEditor($root[0]);
+                faviconEditor.init();
+
+                // Override serializer to return raw image data, we cannot use the default
```
Comment (asmblah) [inline tests/js/web/Notifications/FaviconEditorTest.js]: For consistency, should we also do this for the test above?

File: tests/js/web/Notifications/FaviconEditorTest.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
+const FaviconEditor = require('../../../../js/Notifications/FaviconEditor');
+const $ = require('jquery');
+const vanillaFavicon = require('./snapshots/favicon');
```
Comment (asmblah) [inline tests/js/web/Notifications/FaviconEditorTest.js]: Small thing, but this is more of a "fixture" than a snapshot as you wouldn't expect to need to update the fixture image very often, should this be in a folder called `fixtures` to differentiate them?

File: js/main.js (hunk 1 above)
Comment (mikedevelops) [reply js/main.js]: I'm hoping to move this over to the Lexicon bundle once that is fixed.

File: tests/js/web/Notifications/FaviconEditorTest.js (hunk 6 above)
Comment (mikedevelops) [reply tests/js/web/Notifications/FaviconEditorTest.js]: Yep I agree, I'll change this.

File: tests/js/web/Notifications/FaviconEditorTest.js (hunk 5 above)
Comment (mikedevelops) [reply tests/js/web/Notifications/FaviconEditorTest.js]: Yeah good point, the compression / rendering doesn't seem to differ on the 2 environments when using primitive shapes – I'll change it though, probably best to remove compression from the equation.

File: .babelrc (hunk 2 above)
Comment (mikedevelops) [reply .babelrc]: ... maaaaaybe 😸

File: tests/js/web/Notifications/FaviconEditorTest.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            // not asserting the actual href in this case, as it will be prepended by
+            // the host and port used by Phantom, I'm happy that if the value is not what
+            // I've changed it to then it has been restored
+            $html.find('link').each((index, element) => {
+                expect(element.href).to.not.equal('foo');
+            });
+        });
+    });
+
+    describe('serializer', () => {
```
Comment (asmblah) [inline tests/js/web/Notifications/FaviconEditorTest.js]: Is this page intentionally left blank?

File: docs/_components/favicon.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+ * @param canvas {HTMLCanvasElement}
+ * @param context {CanvasRenderingContext2D}
+ * @param originalImage {HTMLImageElement} derived from the original favicon <link/> node
+ */
+function (canvas, context, originalImage) {
+    return canvas.toDataURL(`image/${getFileExtension(originalImage.src)}`);
+}
+
+// In this example we return the raw image data from our serializer, this return
+// value will be our promise resovle value when invoking update methods
```
Comment (asmblah) [inline docs/_components/favicon.md]: Typo in `resovle`

File: tests/js/web/Notifications/FaviconEditorTest.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
```
Comment (asmblah) [inline tests/js/web/Notifications/FaviconEditorTest.js]: Tests for this method are missing (unless they are somewhere else instead 😄)

**Pull request merged**